File: src/services/ml/recurringDetector.ts
import type { DetectedSubscription, NormalizedTransaction } from "../../types";

interface TransactionGroup {
  description: string;
  merchant?: string;
  transactions: NormalizedTransaction[];
  amounts: number[];
  dates: Date[];
}

export class RecurringDetector {
  private readonly MIN_OCCURRENCES = 2;
  private readonly AMOUNT_VARIANCE_THRESHOLD = 0.1; // 10%
  private readonly MAX_INTERVAL_VARIANCE_DAYS = 5;

  // ─── Known Subscription Merchants ─────────────────────────────────────────
  private readonly KNOWN_SUBSCRIPTIONS = new Map<string, {
    name: string;
    category: string;
    subcategory: string;
    frequency: DetectedSubscription["frequency"];
  }>([
    ["netflix", { name: "Netflix", category: "Subscriptions", subcategory: "Streaming", frequency: "monthly" }],
    ["spotify", { name: "Spotify", category: "Subscriptions", subcategory: "Music", frequency: "monthly" }],
    ["youtube premium", { name: "YouTube Premium", category: "Subscriptions", subcategory: "Streaming", frequency: "monthly" }],
    ["amazon prime", { name: "Amazon Prime", category: "Subscriptions", subcategory: "Streaming", frequency: "monthly" }],
    ["apple music", { name: "Apple Music", category: "Subscriptions", subcategory: "Music", frequency: "monthly" }],
    ["icloud", { name: "iCloud", category: "Subscriptions", subcategory: "Cloud Services", frequency: "monthly" }],
    ["showmax", { name: "Showmax", category: "Subscriptions", subcategory: "Streaming", frequency: "monthly" }],
    ["dstv", { name: "DSTV", category: "Bills", subcategory: "Cable TV", frequency: "monthly" }],
    ["gotv", { name: "GOTV", category: "Bills", subcategory: "Cable TV", frequency: "monthly" }],
    ["microsoft 365", { name: "Microsoft 365", category: "Subscriptions", subcategory: "Software", frequency: "monthly" }],
    ["office 365", { name: "Microsoft 365", category: "Subscriptions", subcategory: "Software", frequency: "monthly" }],
    ["google one", { name: "Google One", category: "Subscriptions", subcategory: "Cloud Services", frequency: "monthly" }],
    ["dropbox", { name: "Dropbox", category: "Subscriptions", subcategory: "Cloud Services", frequency: "monthly" }],
    ["adobe", { name: "Adobe", category: "Subscriptions", subcategory: "Software", frequency: "monthly" }],
    ["canva", { name: "Canva Pro", category: "Subscriptions", subcategory: "Software", frequency: "monthly" }],
    ["coursera", { name: "Coursera", category: "Subscriptions", subcategory: "Education", frequency: "monthly" }],
    ["audible", { name: "Audible", category: "Subscriptions", subcategory: "Education", frequency: "monthly" }],
    ["chatgpt", { name: "ChatGPT Plus", category: "Subscriptions", subcategory: "Software", frequency: "monthly" }],
    ["nordvpn", { name: "NordVPN", category: "Subscriptions", subcategory: "Software", frequency: "monthly" }],
    ["expressvpn", { name: "ExpressVPN", category: "Subscriptions", subcategory: "Software", frequency: "monthly" }],
  ]);

  detect(transactions: NormalizedTransaction[]): DetectedSubscription[] {
    const subscriptions: DetectedSubscription[] = [];

    // Step 1: Check for known subscription merchants
    const knownSubs = this.detectKnownSubscriptions(transactions);
    subscriptions.push(...knownSubs);

    // Step 2: Detect recurring patterns by grouping similar transactions
    const recurringPatterns = this.detectRecurringPatterns(transactions, knownSubs);
    subscriptions.push(...recurringPatterns);

    return this.deduplicateSubscriptions(subscriptions);
  }

  private detectKnownSubscriptions(transactions: NormalizedTransaction[]): DetectedSubscription[] {
    const result: DetectedSubscription[] = [];
    const groupedByMerchant = new Map<string, NormalizedTransaction[]>();

    for (const tx of transactions) {
      const desc = (tx.description || "").toLowerCase();
      const merchant = (tx.merchant || "").toLowerCase();
      const combined = `${desc} ${merchant}`;

      for (const [key, info] of this.KNOWN_SUBSCRIPTIONS) {
        if (combined.includes(key)) {
          const existing = groupedByMerchant.get(key) || [];
          existing.push(tx);
          groupedByMerchant.set(key, existing);
          break;
        }
      }
    }

    for (const [key, txs] of groupedByMerchant) {
      if (txs.length < 1) continue;

      const info = this.KNOWN_SUBSCRIPTIONS.get(key)!;
      const amounts = txs.map((t) => t.amount);
      const dates = txs.map((t) => t.date).sort((a, b) => a.getTime() - b.getTime());
      const avgAmount = amounts.reduce((s, a) => s + a, 0) / amounts.length;

      const sub: DetectedSubscription = {
        name: info.name,
        merchant: info.name,
        amount: avgAmount,
        currency: txs[0].currency || "NGN",
        frequency: info.frequency,
        occurrences: dates,
        firstSeen: dates[0],
        lastSeen: dates[dates.length - 1],
        nextExpected: this.predictNextDate(dates, info.frequency),
        confidence: txs.length >= 2 ? 0.97 : 0.85,
        category: info.category,
        subcategory: info.subcategory,
        annualCost: this.calculateAnnualCost(avgAmount, info.frequency),
      };

      result.push(sub);
    }

    return result;
  }

  private detectRecurringPatterns(
    transactions: NormalizedTransaction[],
    alreadyFound: DetectedSubscription[]
  ): DetectedSubscription[] {
    const result: DetectedSubscription[] = [];
    const foundMerchants = new Set(alreadyFound.map((s) => s.merchant.toLowerCase()));

    // Group transactions by normalized description
    const groups = new Map<string, NormalizedTransaction[]>();

    for (const tx of transactions) {
      if (tx.type !== "debit") continue;

      const key = this.normalizeDescription(tx.description);
      if (key.length < 3) continue;
      if (foundMerchants.has(key)) continue;

      const existing = groups.get(key) || [];
      existing.push(tx);
      groups.set(key, existing);
    }

    for (const [key, txs] of groups) {
      if (txs.length < this.MIN_OCCURRENCES) continue;

      const amounts = txs.map((t) => t.amount);
      const dates = txs.map((t) => t.date).sort((a, b) => a.getTime() - b.getTime());

      if (!this.hasConsistentAmounts(amounts)) continue;

      const frequency = this.detectFrequency(dates);
      if (!frequency) continue;

      const avgAmount = amounts.reduce((s, a) => s + a, 0) / amounts.length;

      result.push({
        name: this.toTitleCase(key),
        merchant: this.toTitleCase(key),
        amount: avgAmount,
        currency: txs[0].currency || "NGN",
        frequency,
        occurrences: dates,
        firstSeen: dates[0],
        lastSeen: dates[dates.length - 1],
        nextExpected: this.predictNextDate(dates, frequency),
        confidence: Math.min(0.5 + txs.length * 0.1, 0.9),
        category: txs[0].category || "Subscriptions",
        subcategory: txs[0].subcategory,
        annualCost: this.calculateAnnualCost(avgAmount, frequency),
      });
    }

    return result;
  }

  private normalizeDescription(description: string): string {
    return description
      .toLowerCase()
      .replace(/\d+/g, "")
      .replace(/[^a-z\s]/g, " ")
      .replace(/\s+/g, " ")
      .trim()
      .slice(0, 30);
  }

  private hasConsistentAmounts(amounts: number[]): boolean {
    if (amounts.length < 2) return false;
    const avg = amounts.reduce((s, a) => s + a, 0) / amounts.length;
    const maxVariance = avg * this.AMOUNT_VARIANCE_THRESHOLD;
    return amounts.every((a) => Math.abs(a - avg) <= maxVariance);
  }

  private detectFrequency(dates: Date[]): DetectedSubscription["frequency"] | null {
    if (dates.length < 2) return null;

    const intervals: number[] = [];
    for (let i = 1; i < dates.length; i++) {
      const diffDays =
        (dates[i].getTime() - dates[i - 1].getTime()) / (1000 * 60 * 60 * 24);
      intervals.push(diffDays);
    }

    const avgInterval = intervals.reduce((s, d) => s + d, 0) / intervals.length;
    const variance =
      intervals.reduce((s, d) => s + Math.abs(d - avgInterval), 0) /
      intervals.length;

    if (variance > this.MAX_INTERVAL_VARIANCE_DAYS) return null;

    if (avgInterval <= 2) return "daily";
    if (avgInterval >= 5 && avgInterval <= 10) return "weekly";
    if (avgInterval >= 25 && avgInterval <= 35) return "monthly";
    if (avgInterval >= 85 && avgInterval <= 95) return "quarterly";
    if (avgInterval >= 355 && avgInterval <= 375) return "annual";

    return null;
  }

  private predictNextDate(
    dates: Date[],
    frequency: DetectedSubscription["frequency"]
  ): Date | undefined {
    if (dates.length === 0) return undefined;
    const last = dates[dates.length - 1];
    const next = new Date(last);

    switch (frequency) {
      case "daily":
        next.setDate(next.getDate() + 1);
        break;
      case "weekly":
        next.setDate(next.getDate() + 7);
        break;
      case "monthly":
        next.setMonth(next.getMonth() + 1);
        break;
      case "quarterly":
        next.setMonth(next.getMonth() + 3);
        break;
      case "annual":
        next.setFullYear(next.getFullYear() + 1);
        break;
    }

    return next;
  }

  private calculateAnnualCost(
    amount: number,
    frequency: DetectedSubscription["frequency"]
  ): number {
    switch (frequency) {
      case "daily":
        return amount * 365;
      case "weekly":
        return amount * 52;
      case "monthly":
        return amount * 12;
      case "quarterly":
        return amount * 4;
      case "annual":
        return amount;
      default:
        return amount * 12;
    }
  }

  private deduplicateSubscriptions(subs: DetectedSubscription[]): DetectedSubscription[] {
    const seen = new Set<string>();
    return subs.filter((s) => {
      const key = s.name.toLowerCase();
      if (seen.has(key)) return false;
      seen.add(key);
      return true;
    });
  }

  private toTitleCase(str: string): string {
    return str
      .split(" ")
      .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
      .join(" ");
  }
}

export const recurringDetector = new RecurringDetector();

File: app/services/ml/recurring_detector.py
from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import date, timedelta
from collections import defaultdict

KNOWN_SUBSCRIPTIONS: dict[str, tuple[str, str]] = {
    "netflix": ("Streaming", "monthly"), "spotify": ("Music", "monthly"),
    "youtube premium": ("Streaming", "monthly"), "showmax": ("Streaming", "monthly"),
    "amazon prime": ("Streaming", "monthly"), "prime video": ("Streaming", "monthly"),
    "apple music": ("Music", "monthly"), "deezer": ("Music", "monthly"),
    "dstv": ("Cable TV", "monthly"), "gotv": ("Cable TV", "monthly"),
    "startimes": ("Cable TV", "monthly"),
    "icloud": ("Cloud Storage", "monthly"), "google one": ("Cloud Storage", "monthly"),
    "dropbox": ("Cloud Storage", "monthly"),
    "microsoft 365": ("Software", "annual"), "office 365": ("Software", "annual"),
    "adobe": ("Software", "monthly"), "canva": ("Software", "monthly"),
    "chatgpt": ("Software", "monthly"), "nordvpn": ("VPN", "monthly"),
    "expressvpn": ("VPN", "monthly"),
    "coursera": ("Education", "monthly"), "udemy": ("Education", "one-time"),
    "audible": ("Audiobooks", "monthly"),
    "piggyvest": ("Savings", "monthly"), "cowrywise": ("Savings", "monthly"),
}

FREQ_DAYS = {"daily": 1, "weekly": 7, "monthly": 30, "quarterly": 91, "annual": 365}
FREQ_ANNUAL_MULTIPLIER = {"daily": 365, "weekly": 52, "monthly": 12, "quarterly": 4, "annual": 1}


@dataclass
class DetectedSubscription:
    name: str
    merchant: str
    amount: float
    frequency: str
    subcategory: str
    last_seen: date
    next_expected: date
    annual_cost: float
    confidence: float
    is_known: bool


class RecurringDetector:
    def detect(self, transactions: list[dict]) -> list[DetectedSubscription]:
        results: list[DetectedSubscription] = []
        seen_merchants: set[str] = set()

        # Group by merchant/description
        groups: dict[str, list[dict]] = defaultdict(list)
        for tx in transactions:
            if tx.get("type") != "debit":
                continue
            key = (tx.get("merchant") or tx.get("description", "")).lower()
            if key:
                groups[key].append(tx)

        for key, txs in groups.items():
            # Known subscription
            for sub_name, (subcat, freq) in KNOWN_SUBSCRIPTIONS.items():
                if sub_name in key and len(txs) >= 1:
                    if sub_name in seen_merchants:
                        continue
                    seen_merchants.add(sub_name)
                    amounts = [float(t["amount"]) for t in txs]
                    avg_amount = sum(amounts) / len(amounts)
                    last = max(t["date"] for t in txs)
                    if isinstance(last, str):
                        from datetime import datetime
                        last = datetime.fromisoformat(last).date()
                    freq_days = FREQ_DAYS.get(freq, 30)
                    next_exp = last + timedelta(days=freq_days)
                    annual = avg_amount * FREQ_ANNUAL_MULTIPLIER.get(freq, 12)
                    results.append(DetectedSubscription(
                        name=sub_name.title(), merchant=sub_name.title(),
                        amount=round(avg_amount, 2), frequency=freq,
                        subcategory=subcat, last_seen=last,
                        next_expected=next_exp, annual_cost=round(annual, 2),
                        confidence=0.97, is_known=True,
                    ))
                    break
            else:
                # Unknown recurring — detect by frequency
                if len(txs) < 2:
                    continue
                dates = sorted(
                    t["date"] if isinstance(t["date"], date) else
                    __import__("datetime").datetime.fromisoformat(str(t["date"])).date()
                    for t in txs
                )
                if len(dates) < 2:
                    continue
                gaps = [(dates[i + 1] - dates[i]).days for i in range(len(dates) - 1)]
                avg_gap = sum(gaps) / len(gaps)
                variance = max(gaps) - min(gaps)

                if variance > 10:
                    continue

                amounts = [float(t["amount"]) for t in txs]
                avg_amount = sum(amounts) / len(amounts)
                amount_variance = max(amounts) - min(amounts)

                if amount_variance / avg_amount > 0.15:
                    continue

                freq = self._gap_to_freq(avg_gap)
                if not freq:
                    continue

                if key in seen_merchants:
                    continue
                seen_merchants.add(key)

                last = dates[-1]
                next_exp = last + timedelta(days=int(avg_gap))
                annual = avg_amount * (365 / avg_gap)

                results.append(DetectedSubscription(
                    name=key.title(), merchant=key.title(),
                    amount=round(avg_amount, 2), frequency=freq,
                    subcategory="Recurring", last_seen=last,
                    next_expected=next_exp, annual_cost=round(annual, 2),
                    confidence=0.78, is_known=False,
                ))

        return results

    def _gap_to_freq(self, days: float) -> str | None:
        if days <= 2:
            return "daily"
        if 5 <= days <= 9:
            return "weekly"
        if 25 <= days <= 35:
            return "monthly"
        if 85 <= days <= 97:
            return "quarterly"
        if 355 <= days <= 375:
            return "annual"
        return None


recurring_detector = RecurringDetector()
